### PR TITLE
feat(LoginStrategy): improve readability of ICS login

### PR DESCRIPTION
**Pull request by jrhender**
Previously we were awaiting cacheServerClient.login in the constructor. As this is an antipattern, changed to storing login promise as property. Then, can check promise before using cacheServerClient.

Also, removed the strategyAddress property as it was essentially equivalent to cacheServerClient.address

### CODE REVIEW

File: lib/LoginStrategy.ts [hunk 1]
```diff
@@ -49,8 +49,8 @@ export class LoginStrategy extends BaseStrategy {
   private readonly ipfsStore: DidStore
   private readonly acceptedRoles: Set<string>
   private readonly privateKey: string
-  private strategyAddress?: string
-  private cacheServerClient?: CacheServerClient
+  private readonly cacheServerClient?: CacheServerClient
+  private readonly cacheServerClientLoginPromise?: Promise<string>
```
Comment (Harasz) [inline lib/LoginStrategy.ts:53]:
```suggestion
    private isCacheServerClientAvailable: boolean
```

File: lib/LoginStrategy.ts [hunk 2]
```diff
@@ -84,7 +84,12 @@ export class LoginStrategy extends BaseStrategy {
       )
     }
```
Comment (Harasz) [inline lib/LoginStrategy.ts:85]:
```suggestion
    }
    this.isCacheServerClientAvailable = false;
```

File: lib/LoginStrategy.ts [hunk 3]
```diff
@@ -84,7 +84,12 @@ export class LoginStrategy extends BaseStrategy {
...
     if (cacheServerUrl && privateKey) {
-      this.initCacheServer(privateKey, cacheServerUrl);
+      this.cacheServerClient = new CacheServerClient({
+        privateKey,
+        provider: this.provider,
+        url: cacheServerUrl,
+      });
+      this.cacheServerClientLoginPromise = this.cacheServerClient.login();
```
Comment (Harasz) [inline lib/LoginStrategy.ts:92]:
```suggestion
      this.cacheServerClient.login().then(() => this.isCacheServerClientAvailable = true);
```

File: lib/LoginStrategy.ts [hunk 4]
```diff
@@ -100,17 +105,6 @@ export class LoginStrategy extends BaseStrategy {
...
-      url: cacheServerUrl,
-    });
-    this.strategyAddress = cacheServerClientInstance.address;
-    await cacheServerClientInstance.login();
-    this.cacheServerClient = cacheServerClientInstance;
-  }
-
   /**
    * @description verifies issuer signature, then check that claim issued
    * no latter then `this.numberOfBlocksBack` and user has enrolled with at
```
Comment (Harasz) [inline lib/LoginStrategy.ts:110]: For line 121
```suggestion
   const didDocument = this.isCacheServerClientAvailable && this.cacheServerClient ? await this.cacheServerClient.getDidDocument(payload.iss) : await this.didResolver.read(payload.iss)
```

File: lib/LoginStrategy.ts [hunk 5]
```diff
@@ -219,8 +213,17 @@ export class LoginStrategy extends BaseStrategy {
...
 
+  /**
+   * Checks the necessary initialization of cache-server has occured
+   * The pattern is described here: https://stackoverflow.com/a/49695053 
+   * @returns 
+   */
+  private async isCacheServerClientAvailable() {
+    return await this.cacheServerClientLoginPromise;
+  }
+
```
Comment (Harasz) [inline lib/LoginStrategy.ts:224]:
```suggestion

```

File: lib/LoginStrategy.ts [hunk 6]
```diff
@@ -219,8 +213,17 @@ export class LoginStrategy extends BaseStrategy {
...
   async getRoleDefinition(namespace: string) : Promise<any> {
-    if (this.cacheServerClient) {
+    if (this.cacheServerClient && await this.isCacheServerClientAvailable()) {
```
Comment (Harasz) [inline lib/LoginStrategy.ts:226]:
```suggestion
    if (this.cacheServerClient && this.isCacheServerClientAvailable) {
```

File: lib/LoginStrategy.ts [hunk 7]
```diff
@@ -230,7 +233,7 @@ export class LoginStrategy extends BaseStrategy {
   }
 
   async getUserClaims(did: string): Promise<Claim[]> {
-    if (this.cacheServerClient) {
+    if (this.cacheServerClient && await this.cacheServerClientLoginPromise) {
```
Comment (Harasz) [inline lib/LoginStrategy.ts:236]:
```suggestion
    if (this.cacheServerClient && this.isCacheServerClientAvailable) {
```

**Review by Harasz — COMMENTED** [review]
@jrhender It looks a way better but it doesn't work with https://github.com/energywebfoundation/iam-cache-server/pull/174/files. I manged to fix (check my suggestions). The only thing I don't like in my suggestions is `this.cacheServerClient.login().then(() => this.isCacheServerClientAvailable = true);` We should find a better replacement for this.

**Comment by jrhender** [comment]
@Harasz Thanks for the review and feedback. How does it look to you now?

**Review by Harasz — APPROVED** [review]
@jrhender I checked it and everything works fine. I wonder if we can find a better replacement for `.then()` in constructor.

**Comment by jrhender** [comment]
@Harasz what do you not like about `.then()` in constructor? I think our options to not use `.then()` in constructor are:
1. use async function as you are currently doing on `task/ICS-62-fix-cache-server-logging` branch
2. have the promise as a property and await (as I was doing previously in this PR)

I don't like 1. because I think it is confusing to be using an `async` function without `awaiting` it. It's just quite an irregular pattern. I think using `.then()` is a much more standard approach
I think 2. is fine, it arguably makes the constructor a bit cleaner, but might be a bit more confusing than just using `.then()`

**Pull request merged**